src/shell/types.rs

```diff
@@ -632,6 +658,81 @@ impl KillSignalDropGuard {
   }
 }
 
+#[derive(Debug, Default)]
+struct ProcessSignalerInner {
+  /// The PID of the current foreground process, if any.
+  current_pid: Cell<Option<u32>>,
+  /// Sender for process spawn notifications.
+  /// Lazily initialized on first subscribe.
+  sender: RefCell<Option<broadcast::Sender<u32>>>,
+}
+
+/// Provides access to the currently running foreground child process.
+///
+/// This is useful for signal forwarding scenarios where you need to check
+/// if the child process is in the same process group as the parent.
+///
+/// # Example
+///
+/// ```ignore
+/// let signaler = ProcessSignaler::new();
+/// let mut receiver = signaler.subscribe();
+///
+/// // In a signal handler, check if we should forward the signal
+/// if let Some(child_pid) = signaler.current_pid() {
+///     // Check if child is in same process group
+///     let child_pgid = unsafe { libc::getpgid(child_pid as i32) };
+///     let our_pgid = unsafe { libc::getpgid(0) };
+///
+///     if child_pgid != our_pgid {
+///         // Child in different process group, forward signal
+///         kill_signal.send(SignalKind::SIGINT);
+///     }
+/// }
+/// ```
+#[derive(Debug, Clone, Default)]
+pub struct ProcessSignaler(Rc<ProcessSignalerInner>);
+
+impl ProcessSignaler {
+  /// Creates a new ProcessSignaler.
+  pub fn new() -> Self {
+    Self::default()
+  }
+
+  /// Returns the PID of the current foreground child process, if any.
+  ///
+  /// Returns `None` if no child process is currently running.
+  pub fn current_pid(&self) -> Option<u32> {
```

Comment:
There could be multiple child processes running at the same time:

```sh
deno eval 'console.log(1)' ; deno eval 'console.log(2)' &
```

It seems this would only hold one of them?